wsf_scraping/spiders/unicef_spider.py

```diff
@@ -0,0 +1,119 @@
+import scrapy
+from urllib.parse import urlparse
+from scrapy.http import Request
+from wsf_scraping.items import UNICEFArticle
+from scrapy.spidermiddlewares.httperror import HttpError
+from twisted.internet.error import DNSLookupError
+from twisted.internet.error import TimeoutError
+
+
+class UnicefSpider(scrapy.Spider):
+    name = 'unicef'
+
+    custom_settings = {
+        'JOBDIR': 'crawls/unicef'
+    }
+
+    def on_error(self, failure):
+        self.logger.error(repr(failure))
+
+        if failure.check(HttpError):
+            response = failure.value.response
+            self.logger.error(
+                'HttpError on %s (%s)',
+                response.url,
+                response.status,
+            )
+
+        elif failure.check(DNSLookupError):
+            request = failure.request
+            self.logger.error('DNSLookupError on %s', request.url)
+
+        elif failure.check(TimeoutError):
+            request = failure.request
+            self.logger.error('TimeoutError on %s', request.url)
+
+    def start_requests(self):
+        """ This sets up the urls to scrape for each years.
+        """
+
+        urls = [
+            'https://data.unicef.org/resources/resource-type/publication/',
+            'https://data.unicef.org/resources/resource-type/guidance/'
+        ]
+
+        for url in urls:
+            self.logger.info('Initial url: %s', url)
+            yield scrapy.Request(
+                url=url,
+                callback=self.parse,
+                errback=self.on_error,
+                dont_filter=True,
+            )
+
+    def parse(self, response):
+        """ Parse the articles listing page and go to the next one.
+
+        @url https://data.unicef.org/resources/resource-type/publication/
+        @returns items 0 0
+        @returns requests 115
```

Comment:
I don't understand what this docstring means. It takes a response, not a URL – and what is 0 0?

Reply:
It's the way scrapy deals with unit testing. `url` is the test url to scrape, `returns items` means it will returns between 0 and 0 items, and `returns requests 115` means it will returns at least 115 requests (it's the actual amount of matching links on the page here).

It's not a very clear way to test the code, but it seems to be the better one with scrapy at the moment...